loader.go

```diff
@@ -345,24 +346,32 @@ func loadPrograms(elfFile *elf.File, maps map[string]Map) (map[string]Program, e
 			}
 		}
 
-		// One section may contain multiple programs
-		// Cut section's bytecode into programs based on symbols information
-		lastOffset := len(bytecode)
-		for i := len(symbols) - 1; i >= 0; i-- {
-			// Skip symbols which are either:
-			// 1. non GLOBAL binded
-			// 2. Don't belong to current section
-			symbol := &symbols[i]
-			if int(symbol.Section) != sectionIndex || elf.ST_BIND(symbol.Info) != elf.STB_GLOBAL {
-				continue
+		// One section may contain multiple programs.
+		// Find all programs and it offsets from symbols table, then
+		// reverse sort them by offset (since order is not guaranteed!)
+		offsetToNameMap := map[int]string{}
+		offsetToNameKeys := []int{} // For keys sort
+		for _, symbol := range symbols {
+			if int(symbol.Section) == sectionIndex && elf.ST_BIND(symbol.Info) == elf.STB_GLOBAL {
+				key := int(symbol.Value)
+				offsetToNameMap[key] = symbol.Name
+				offsetToNameKeys = append(offsetToNameKeys, key)
 			}
-			offset := int(symbol.Value)
+			// Skip others
+		}
+
+		// Slice eBPF programs by reverse sorted offsets from symbol table
+		sort.Sort(sort.Reverse(sort.IntSlice(offsetToNameKeys)))
+		lastOffset := len(bytecode)
+		for _, offset := range offsetToNameKeys {
+			name := offsetToNameMap[offset]
 			size := lastOffset - offset
+			// Check program size
 			if size/bpfInstructionLen > bpfMaxInstructions {
```

Comment:
should we do this? kernel verifier would throw an error if there are too many instructions anyway. but e.g. you wont need to do a special gating based on kernel's version (recent ones , 5.3+ allow up to 1mil instructions)

Reply:
As I remember kernel's message wasn't too user friendly so I added more useful message, but we can for sure remove it, np here.